core/libp2p/protocol_muxer/multiselect/message_reader.cpp

```diff
@@ -30,73 +31,49 @@ namespace libp2p::protocol_muxer {
-                    const std::error_code &ec, size_t n) mutable {
-                  if (ec || n != 1) {
-                    auto multiselect = connection_state->multiselect_;
-                    multiselect->onError(
-                        std::move(connection_state),
-                        "cannot read from the connection: " + ec.message(), ec);
-                    return;
-                  }
-                  onReadVarintCompleted(std::move(connection_state));
-                });
-  }
-
-  void MessageReader::onReadVarintCompleted(
-      std::shared_ptr<ConnectionState> connection_state) {
-    auto varint_opt = getVarint(*connection_state->read_buffer_);
-    if (!varint_opt) {
-      // no varint; continue reading
-      readNextVarint(std::move(connection_state));
-      return;
+    // we don't know a length of the varint, so gather it byte-by-byte
+    std::vector<uint8_t> varint_bytes{};
+    std::optional<UVarint> varint{std::nullopt};
+    while (!varint) {
+      auto varint_candidate_res = connection_state->read(1);
+      if (!varint_candidate_res) {
+        // something bad happened during read
+        processError(std::move(connection_state), varint_candidate_res.error());
+        return;
+      }
+      varint_bytes.push_back(varint_candidate_res.value()[0]);
+      varint = getVarint(varint_bytes);
     }
-    // we have length of the line to be read; do it
-    connection_state->read_buffer_->consume(varint_opt->size());
-    auto bytes_to_read = varint_opt->toUInt64();
-    readNextBytes(std::move(connection_state), bytes_to_read,
-                  [bytes_to_read](std::shared_ptr<ConnectionState> state) {
-                    onReadLineCompleted(std::move(state), bytes_to_read);
-                  });
+
+    // that varint tells us, how much bytes to read next; do it
+    readNextBytes(std::move(connection_state), varint->toUInt64(),
+                  onReadLineCompleted);
   }
 
   void MessageReader::readNextBytes(
       std::shared_ptr<ConnectionState> connection_state, uint64_t bytes_to_read,
-      std::function<void(std::shared_ptr<ConnectionState>)> final_callback) {
-    auto state = connection_state;
-    state->read(bytes_to_read,
-                [connection_state = std::move(connection_state), bytes_to_read,
-                 final_callback = std::move(final_callback)](
-                    const std::error_code &ec, size_t n) mutable {
-                  if (ec || n != bytes_to_read) {
-                    auto multiselect = connection_state->multiselect_;
-                    multiselect->onError(
-                        std::move(connection_state),
-                        "cannot read from the connection: " + ec.message(), ec);
-                    return;
-                  }
-                  final_callback(std::move(connection_state));
-                });
+      const std::function<void(std::shared_ptr<ConnectionState>, Buffer)>
```

Comment:
Maybe a type alias would make it neater?

Reply:
Well, it's used in just one place and isn't even exposed to some external entities, so I am not sure alias is needed